File: js/src/cfb.ts
/**
 * Microsoft Compound File Binary (OLE2) containers — read and write.
 *
 * Enough of [MS-CFB] v3 (512-byte sectors, 64-byte mini sectors, 4096-byte
 * mini-stream cutoff, DIFAT chains) to produce files Microsoft Project, Apache
 * POI and olefile read, and to read the ones they write. Written from the
 * public [MS-CFB] specification; a direct port of the Python implementation, so
 * both produce the same bytes.
 */

export const FREESECT = 0xffffffff;
export const ENDOFCHAIN = 0xfffffffe;
export const FATSECT = 0xfffffffd;
export const DIFSECT = 0xfffffffc;
export const NOSTREAM = 0xffffffff;

export const SECTOR = 512;
export const MINI_SECTOR = 64;
export const MINI_CUTOFF = 4096;
const FAT_PER_SECTOR = SECTOR / 4;
const DIFAT_PER_SECTOR = FAT_PER_SECTOR - 1; // the last dword chains onward

export const TYPE_STORAGE = 1;
export const TYPE_STREAM = 2;
export const TYPE_ROOT = 5;

const MAGIC = Uint8Array.from([0xd0, 0xcf, 0x11, 0xe0, 0xa1, 0xb1, 0x1a, 0xe1]);

/** A storage node: an ordered tree of child storages and streams. */
export class Storage {
  /** Insertion-ordered, and a Map because JS objects reorder integer-like keys. */
  readonly children = new Map<string, Storage | Uint8Array>();

  addStream(name: string, data: Uint8Array): void {
    this.children.set(name, data);
  }

  addStorage(name: string): Storage {
    const existing = this.children.get(name);
    if (existing instanceof Storage) return existing;
    const s = new Storage();
    this.children.set(name, s);
    return s;
  }

  /** `set("a/b/c", data)` creating intermediate storages. */
  set(path: string, data: Uint8Array): void {
    const parts = path.split("/");
    let s: Storage = this;
    for (const p of parts.slice(0, -1)) s = s.addStorage(p);
    s.addStream(parts[parts.length - 1]!, data);
  }

  /** The stream at `path`, or undefined when it is absent or a storage. */
  get(path: string): Uint8Array | undefined {
    const node = this.node(path);
    return node instanceof Uint8Array ? node : undefined;
  }

  storage(path: string): Storage {
    let s: Storage = this;
    for (const p of path.split("/")) s = s.addStorage(p);
    return s;
  }

  private node(path: string): Storage | Uint8Array | undefined {
    let cur: Storage | Uint8Array | undefined = this;
    for (const p of path.split("/")) {
      if (!(cur instanceof Storage)) return undefined;
      cur = cur.children.get(p);
    }
    return cur;
  }

  /** Every stream path in the tree, depth first. */
  paths(prefix = ""): string[] {
    const out: string[] = [];
    for (const [name, val] of this.children) {
      const p = prefix ? `${prefix}/${name}` : name;
      if (val instanceof Storage) out.push(...val.paths(p));
      else out.push(p);
    }
    return out;
  }
}

class Entry {
  child = NOSTREAM;
  left = NOSTREAM;
  right = NOSTREAM;
  start = ENDOFCHAIN;
  size = 0;
  index = -1;
  clsid = new Uint8Array(16);
  name: string;
  etype: number;
  data: Uint8Array | undefined;

  // plain assignment, not parameter properties: Node runs this file by
  // stripping types, which only supports erasable syntax
  constructor(name: string, etype: number, data?: Uint8Array) {
    this.name = name;
    this.etype = etype;
    this.data = data;
  }
}

function utf16le(name: string): Uint8Array {
  const out = new Uint8Array(name.length * 2);
  const view = new DataView(out.buffer);
  for (let i = 0; i < name.length; i++) view.setUint16(i * 2, name.charCodeAt(i), true);
  return out;
}

function fromUtf16le(bytes: Uint8Array): string {
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  let s = "";
  for (let i = 0; i + 1 < bytes.byteLength; i += 2) s += String.fromCharCode(view.getUint16(i, true));
  return s;
}

/**
 * [MS-CFB] 2.6.4 ordering: by UTF-16 name length first, then upper-cased code
 * units. Exported because getting it wrong makes readers lose entries.
 */
export function compareNames(a: string, b: string): number {
  if (a.length !== b.length) return a.length - b.length;
  const ua = a.toUpperCase();
  const ub = b.toUpperCase();
  for (let i = 0; i < ua.length; i++) {
    const d = ua.charCodeAt(i) - ub.charCodeAt(i);
    if (d !== 0) return d;
  }
  return 0;
}

function buildTree(entries: Entry[]): number {
  if (entries.length === 0) return NOSTREAM;
  const mid = Math.floor(entries.length / 2);
  const node = entries[mid]!;
  node.left = buildTree(entries.slice(0, mid));
  node.right = buildTree(entries.slice(mid + 1));
  return node.index;
}

function pad(data: Uint8Array, unit: number): Uint8Array {
  const rem = data.length % unit;
  if (rem === 0) return data;
  const out = new Uint8Array(data.length + (unit - rem));
  out.set(data);
  return out;
}

function dirEntry(e: Entry | null): Uint8Array {
  const buf = new Uint8Array(128);
  const view = new DataView(buf.buffer);
  if (e === null) {
    view.setUint32(68, NOSTREAM, true);
    view.setUint32(72, NOSTREAM, true);
    view.setUint32(76, NOSTREAM, true);
    return buf;
  }
  const name = utf16le(e.name);
  if (name.length + 2 > 64) throw new Error(`name too long: ${e.name}`);
  buf.set(name, 0);
  view.setUint16(64, name.length + 2, true); // includes the terminating NUL
  buf[66] = e.etype;
  buf[67] = 1; // colour: black
  view.setUint32(68, e.left, true);
  view.setUint32(72, e.right, true);
  view.setUint32(76, e.child, true);
  buf.set(e.clsid, 80);
  const start = e.start !== ENDOFCHAIN ? e.start : e.etype !== TYPE_STREAM ? 0 : ENDOFCHAIN;
  view.setUint32(116, start, true);
  view.setBigUint64(120, BigInt(e.size), true);
  return buf;
}

/** Serialise a storage tree into a compound file. */
export function writeCfb(root: Storage, rootClsid = new Uint8Array(16)): Uint8Array {
  const entries: Entry[] = [];
  const rootEntry = new Entry("Root Entry", TYPE_ROOT);
  rootEntry.clsid = rootClsid;
  entries.push(rootEntry);

  const flatten = (storage: Storage, parent: Entry): void => {
    const kids: Entry[] = [];
    for (const [name, val] of storage.children) {
      const e =
        val instanceof Storage ? new Entry(name, TYPE_STORAGE) : new Entry(name, TYPE_STREAM, val);
      entries.push(e);
      e.index = entries.length - 1;
      kids.push(e);
    }
    kids.sort((a, b) => compareNames(a.name, b.name));
    parent.child = buildTree(kids);
    for (const e of kids) {
      if (e.etype === TYPE_STORAGE) flatten(storage.children.get(e.name) as Storage, e);
    }
  };
  rootEntry.index = 0;
  flatten(root, rootEntry);
  entries.forEach((e, i) => (e.index = i));

  // mini stream: every stream below the cutoff
  const miniParts: Uint8Array[] = [];
  const minifat: number[] = [];
  let miniLen = 0;
  for (const e of entries) {
    if (e.etype === TYPE_STREAM && e.data && e.data.length > 0 && e.data.length < MINI_CUTOFF) {
      e.size = e.data.length;
      e.start = minifat.length;
      const padded = pad(e.data, MINI_SECTOR);
      const n = padded.length / MINI_SECTOR;
      miniParts.push(padded);
      miniLen += padded.length;
      for (let i = 1; i < n; i++) minifat.push(minifat.length + 1);
      minifat.push(ENDOFCHAIN);
    }
  }
  const miniStream = new Uint8Array(miniLen);
  {
    let off = 0;
    for (const p of miniParts) {
      miniStream.set(p, off);
      off += p.length;
    }
  }
  rootEntry.size = miniStream.length;

  const sectors: Uint8Array[] = [];
  const fat: number[] = [];
  const addChain = (data: Uint8Array): number => {
    const padded = pad(data, SECTOR);
    const n = padded.length / SECTOR;
    if (n === 0) return ENDOFCHAIN;
    const first = sectors.length;
    for (let i = 0; i < n; i++) {
      sectors.push(padded.subarray(i * SECTOR, (i + 1) * SECTOR));
      fat.push(i < n - 1 ? first + i + 1 : ENDOFCHAIN);
    }
    return first;
  };

  for (const e of entries) {
    if (e.etype === TYPE_STREAM && e.data && e.data.length >= MINI_CUTOFF) {
      e.size = e.data.length;
      e.start = addChain(e.data);
    }
  }
  rootEntry.start = miniStream.length ? addChain(miniStream) : ENDOFCHAIN;

  let minifatStart = ENDOFCHAIN;
  let minifatCount = 0;
  if (minifat.length) {
    const raw = pad(new Uint8Array(minifat.length * 4), SECTOR);
    const view = new DataView(raw.buffer);
    minifat.forEach((v, i) => view.setUint32(i * 4, v, true));
    raw.fill(0xff, minifat.length * 4); // unused slots are FREESECT
    minifatStart = addChain(raw);
    minifatCount = raw.length / SECTOR;
  }

  // directory, padded out with valid empty entries
  const dirEntries: Uint8Array[] = entries.map((e) => dirEntry(e));
  const slotBytes = pad(new Uint8Array(dirEntries.length * 128), SECTOR).length;
  while (dirEntries.length * 128 < slotBytes) dirEntries.push(dirEntry(null));
  const dirRaw = new Uint8Array(dirEntries.length * 128);
  dirEntries.forEach((d, i) => dirRaw.set(d, i * 128));
  const dirStart = addChain(dirRaw);

  // FAT and DIFAT sectors need FAT entries of their own, so iterate to a fixpoint
  const nData = sectors.length;
  let nFat = 0;
  let nDifat = 0;
  for (;;) {
    const neededFat = Math.ceil((nData + nFat + nDifat) / FAT_PER_SECTOR);
    const neededDifat = Math.ceil(Math.max(0, neededFat - 109) / DIFAT_PER_SECTOR);
    if (neededFat === nFat && neededDifat === nDifat) break;
    nFat = neededFat;
    nDifat = neededDifat;
  }

  const fatFull = [...fat];
  for (let i = 0; i < nFat; i++) fatFull.push(FATSECT);
  for (let i = 0; i < nDifat; i++) fatFull.push(DIFSECT);
  while (fatFull.length < nFat * FAT_PER_SECTOR) fatFull.push(FREESECT);
  const fatRaw = new Uint8Array(fatFull.length * 4);
  {
    const view = new DataView(fatRaw.buffer);
    fatFull.forEach((v, i) => view.setUint32(i * 4, v, true));
  }
  const fatSectorIds = Array.from({ length: nFat }, (_, i) => nData + i);
  const difatSectorIds = Array.from({ length: nDifat }, (_, i) => nData + nFat + i);
  for (let i = 0; i < nFat; i++) sectors.push(fatRaw.subarray(i * SECTOR, (i + 1) * SECTOR));
  for (let i = 0; i < nDifat; i++) {
    const chunk = fatSectorIds.slice(109 + i * DIFAT_PER_SECTOR, 109 + (i + 1) * DIFAT_PER_SECTOR);
    const sector = new Uint8Array(SECTOR);
    const view = new DataView(sector.buffer);
    for (let j = 0; j < DIFAT_PER_SECTOR; j++) view.setUint32(j * 4, chunk[j] ?? FREESECT, true);
    view.setUint32(SECTOR - 4, i + 1 < nDifat ? difatSectorIds[i + 1]! : ENDOFCHAIN, true);
    sectors.push(sector);
  }

  const header = new Uint8Array(SECTOR);
  const hv = new DataView(header.buffer);
  header.set(MAGIC, 0);
  hv.setUint16(24, 0x003e, true); // minor version
  hv.setUint16(26, 0x0003, true); // major version 3
  hv.setUint16(28, 0xfffe, true); // little endian
  hv.setUint16(30, 9, true); // 512-byte sectors
  hv.setUint16(32, 6, true); // 64-byte mini sectors
  hv.setUint32(40, 0, true); // directory sector count (v3: 0)
  hv.setUint32(44, nFat, true);
  hv.setUint32(48, dirStart, true);
  hv.setUint32(56, MINI_CUTOFF, true);
  hv.setUint32(60, minifatStart, true);
  hv.setUint32(64, minifatCount, true);
  hv.setUint32(68, nDifat ? difatSectorIds[0]! : ENDOFCHAIN, true);
  hv.setUint32(72, nDifat, true);
  for (let i = 0; i < 109; i++) hv.setUint32(76 + i * 4, fatSectorIds[i] ?? FREESECT, true);

  const out = new Uint8Array(SECTOR + sectors.length * SECTOR);
  out.set(header, 0);
  sectors.forEach((s, i) => out.set(s, SECTOR + i * SECTOR));
  return out;
}

/** Parse a compound file into a storage tree. */
export function readCfb(bytes: Uint8Array): Storage {
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  for (let i = 0; i < MAGIC.length; i++) {
    if (bytes[i] !== MAGIC[i]) throw new Error("not a compound file (bad magic)");
  }
  const sectorSize = 1 << view.getUint16(30, true);
  const miniSectorSize = 1 << view.getUint16(32, true);
  const nFat = view.getUint32(44, true);
  const dirStart = view.getUint32(48, true);
  const miniCutoff = view.getUint32(56, true);
  const minifatStart = view.getUint32(60, true);
  const nDifat = view.getUint32(72, true);
  let difatStart = view.getUint32(68, true);

  const sectorOffset = (id: number) => sectorSize + id * sectorSize;
  const readSector = (id: number) => bytes.subarray(sectorOffset(id), sectorOffset(id) + sectorSize);

  // the DIFAT lists the FAT sectors: 109 in the header, the rest chained
  const fatSectorIds: number[] = [];
  for (let i = 0; i < Math.min(nFat, 109); i++) fatSectorIds.push(view.getUint32(76 + i * 4, true));
  for (let n = 0; n < nDifat && difatStart < FATSECT; n++) {
    const sec = readSector(difatStart);
    const sv = new DataView(sec.buffer, sec.byteOffset, sec.byteLength);
    for (let i = 0; i < sectorSize / 4 - 1; i++) {
      const id = sv.getUint32(i * 4, true);
      if (id < FATSECT) fatSectorIds.push(id);
    }
    difatStart = sv.getUint32(sectorSize - 4, true);
  }

  const fat: number[] = [];
  for (const id of fatSectorIds) {
    const sec = readSector(id);
    const sv = new DataView(sec.buffer, sec.byteOffset, sec.byteLength);
    for (let i = 0; i < sectorSize / 4; i++) fat.push(sv.getUint32(i * 4, true));
  }

  const chain = (start: number): number[] => {
    const out: number[] = [];
    let s = start;
    const seen = new Set<number>();
    while (s < FATSECT && !seen.has(s)) {
      seen.add(s);
      out.push(s);
      s = fat[s] ?? ENDOFCHAIN;
    }
    return out;
  };
  const readChain = (start: number, size?: number): Uint8Array => {
    const ids = chain(start);
    const out = new Uint8Array(ids.length * sectorSize);
    ids.forEach((id, i) => out.set(readSector(id), i * sectorSize));
    return size === undefined ? out : out.subarray(0, size);
  };

  const dirData = readChain(dirStart);
  const nEntries = Math.floor(dirData.length / 128);
  type Raw = { name: string; etype: number; child: number; left: number; right: number; start: number; size: number };
  const raw: Raw[] = [];
  for (let i = 0; i < nEntries; i++) {
    const e = dirData.subarray(i * 128, (i + 1) * 128);
    const ev = new DataView(e.buffer, e.byteOffset, e.byteLength);
    const nameLen = ev.getUint16(64, true);
    raw.push({
      name: nameLen > 2 ? fromUtf16le(e.subarray(0, nameLen - 2)) : "",
      etype: e[66]!,
      child: ev.getUint32(76, true),
      left: ev.getUint32(68, true),
      right: ev.getUint32(72, true),
      start: ev.getUint32(116, true),
      size: Number(ev.getBigUint64(120, true)),
    });
  }

  const rootRaw = raw[0];
  if (!rootRaw) throw new Error("compound file has no root entry");
  const miniStream = rootRaw.start < FATSECT ? readChain(rootRaw.start, rootRaw.size) : new Uint8Array(0);
  const minifat: number[] = [];
  if (minifatStart < FATSECT) {
    const mf = readChain(minifatStart);
    const mv = new DataView(mf.buffer, mf.byteOffset, mf.byteLength);
    for (let i = 0; i < mf.length / 4; i++) minifat.push(mv.getUint32(i * 4, true));
  }
  const readMini = (start: number, size: number): Uint8Array => {
    const out = new Uint8Array(size);
    let s = start;
    let off = 0;
    const seen = new Set<number>();
    while (s < FATSECT && off < size && !seen.has(s)) {
      seen.add(s);
      const take = Math.min(miniSectorSize, size - off);
      out.set(miniStream.subarray(s * miniSectorSize, s * miniSectorSize + take), off);
      off += take;
      s = minifat[s] ?? ENDOFCHAIN;
    }
    return out;
  };

  const root = new Storage();
  const walk = (index: number, into: Storage): void => {
    if (index === NOSTREAM || index >= raw.length) return;
    const e = raw[index]!;
    walk(e.left, into);
    if (e.etype === TYPE_STORAGE) {
      walk(e.child, into.addStorage(e.name));
    } else if (e.etype === TYPE_STREAM) {
      const data =
        e.size === 0
          ? new Uint8Array(0)
          : e.size < miniCutoff
            ? readMini(e.start, e.size)
            : readChain(e.start, e.size);
      into.addStream(e.name, data);
    }
    walk(e.right, into);
  };
  walk(rootRaw.child, root);
  sortChildren(root);
  return root;
}

/**
 * Order every storage's children by name.
 *
 * The directory tree's shape is not semantic — two readers can walk the same
 * file into differently ordered trees — but the writer emits entries in
 * insertion order, so the order decides the output bytes. Canonicalising here
 * keeps this implementation and the Python one producing identical files from
 * identical input, which is what the parity tests check.
 */
function sortChildren(node: Storage): void {
  const sorted = [...node.children.entries()].sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
  node.children.clear();
  for (const [name, val] of sorted) {
    node.children.set(name, val);
    if (val instanceof Storage) sortChildren(val);
  }
}

File: pymppwriter/cfb.py
"""Minimal writer for Microsoft Compound File Binary (OLE2) containers.

Implements enough of [MS-CFB] v3 (512-byte sectors, 64-byte mini sectors,
4096-byte mini-stream cutoff) to produce files that Apache POI / olefile /
MS Project can read. Written from the public [MS-CFB] specification.
"""
from __future__ import annotations
import struct
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Union

FREESECT = 0xFFFFFFFF
ENDOFCHAIN = 0xFFFFFFFE
FATSECT = 0xFFFFFFFD
DIFSECT = 0xFFFFFFFC
NOSTREAM = 0xFFFFFFFF

SECTOR = 512
MINI_SECTOR = 64
MINI_CUTOFF = 4096
FAT_PER_SECTOR = SECTOR // 4

TYPE_STORAGE, TYPE_STREAM, TYPE_ROOT = 1, 2, 5


@dataclass
class Storage:
    children: Dict[str, Union["Storage", bytes]] = field(default_factory=dict)

    def add_stream(self, name: str, data: bytes) -> None:
        self.children[name] = bytes(data)

    def add_storage(self, name: str) -> "Storage":
        s = self.children.get(name)
        if not isinstance(s, Storage):
            s = Storage()
            self.children[name] = s
        return s

    def set_path(self, path: str, data: bytes) -> None:
        parts = path.split("/")
        s = self
        for p in parts[:-1]:
            s = s.add_storage(p)
        s.add_stream(parts[-1], data)

    def storage_path(self, path: str) -> "Storage":
        s = self
        for p in path.split("/"):
            s = s.add_storage(p)
        return s


class _Entry:
    __slots__ = ("name", "etype", "data", "child", "left", "right", "start", "size", "index", "clsid")

    def __init__(self, name: str, etype: int, data: Optional[bytes] = None):
        self.name = name
        self.etype = etype
        self.data = data
        self.child = NOSTREAM
        self.left = NOSTREAM
        self.right = NOSTREAM
        self.start = ENDOFCHAIN
        self.size = 0
        self.index = -1
        self.clsid = b"\0" * 16


def _name_key(name: str):
    # [MS-CFB] 2.6.4: compare by UTF-16 length first, then upper-cased code units.
    u = name.encode("utf-16-le")
    return (len(u), name.upper().encode("utf-16-le"))


def _build_tree(entries: List[_Entry]) -> int:
    """Return index of root of a balanced binary tree over sorted entries."""
    if not entries:
        return NOSTREAM
    mid = len(entries) // 2
    node = entries[mid]
    node.left = _build_tree(entries[:mid])
    node.right = _build_tree(entries[mid + 1:])
    return node.index


def _pad(data: bytes, unit: int) -> bytes:
    rem = len(data) % unit
    return data if rem == 0 else data + b"\0" * (unit - rem)


def write_cfb(root: Storage, root_clsid: bytes = b"\0" * 16) -> bytes:
    # ---- flatten directory -------------------------------------------------
    entries: List[_Entry] = []
    root_entry = _Entry("Root Entry", TYPE_ROOT)
    root_entry.clsid = root_clsid
    entries.append(root_entry)

    def flatten(storage: Storage, parent: _Entry) -> None:
        kids: List[_Entry] = []
        for name, val in storage.children.items():
            if isinstance(val, Storage):
                e = _Entry(name, TYPE_STORAGE)
            else:
                e = _Entry(name, TYPE_STREAM, val)
            entries.append(e)
            kids.append(e)
        for e in kids:
            e.index = entries.index(e)
        kids.sort(key=lambda e: _name_key(e.name))
        parent.child = _build_tree(kids)
        for e in kids:
            if e.etype == TYPE_STORAGE:
                flatten(storage.children[e.name], e)

    root_entry.index = 0
    flatten(root, root_entry)
    for i, e in enumerate(entries):
        e.index = i

    # ---- mini stream -------------------------------------------------------
    mini_stream = bytearray()
    minifat: List[int] = []
    for e in entries:
        if e.etype == TYPE_STREAM and 0 < len(e.data) < MINI_CUTOFF:
            e.size = len(e.data)
            e.start = len(minifat)
            padded = _pad(e.data, MINI_SECTOR)
            n = len(padded) // MINI_SECTOR
            mini_stream += padded
            minifat.extend(range(len(minifat) + 1, len(minifat) + n))
            minifat.append(ENDOFCHAIN)
    root_entry.size = len(mini_stream)

    # ---- regular sectors ---------------------------------------------------
    sectors: List[bytes] = []          # sector payloads (512 bytes each)
    fat: List[int] = []                # FAT entry per sector

    def add_chain(data: bytes) -> int:
        padded = _pad(data, SECTOR)
        n = len(padded) // SECTOR
        if n == 0:
            return ENDOFCHAIN
        first = len(sectors)
        for i in range(n):
            sectors.append(padded[i * SECTOR:(i + 1) * SECTOR])
            fat.append(first + i + 1 if i < n - 1 else ENDOFCHAIN)
        return first

    # large streams
    for e in entries:
        if e.etype == TYPE_STREAM and len(e.data) >= MINI_CUTOFF:
            e.size = len(e.data)
            e.start = add_chain(e.data)

    # mini stream itself lives in regular sectors, pointed to by root entry
    root_entry.start = add_chain(bytes(mini_stream)) if mini_stream else ENDOFCHAIN

    # minifat
    minifat_start, minifat_count = ENDOFCHAIN, 0
    if minifat:
        raw = b"".join(struct.pack("<I", v) for v in minifat)
        raw = _pad(raw, SECTOR)
        # pad unused minifat slots with FREESECT
        raw = raw[: len(minifat) * 4] + b"\xff" * (len(raw) - len(minifat) * 4)
        minifat_start = add_chain(raw)
        minifat_count = len(raw) // SECTOR

    # directory
    dir_raw = bytearray()
    for e in entries:
        dir_raw += _dir_entry(e)
    dir_raw = _pad(bytes(dir_raw), SECTOR)
    # fill trailing unused entries with valid empty entries
    n_slots = len(dir_raw) // 128
    for i in range(len(entries), n_slots):
        dir_raw = dir_raw[: i * 128] + _dir_entry(None) + dir_raw[(i + 1) * 128:]
    dir_start = add_chain(bytes(dir_raw))

    # FAT sectors (iterate: FAT sectors — and any DIFAT sectors listing them
    # beyond the header's 109 slots — need FAT entries too)
    DIFAT_PER_SECTOR = FAT_PER_SECTOR - 1      # last dword chains to the next DIFAT sector
    n_data = len(sectors)
    n_fat = n_difat = 0
    while True:
        needed_fat = -(-(n_data + n_fat + n_difat) // FAT_PER_SECTOR)
        needed_difat = -(-max(0, needed_fat - 109) // DIFAT_PER_SECTOR)
        if (needed_fat, needed_difat) == (n_fat, n_difat):
            break
        n_fat, n_difat = needed_fat, needed_difat

    fat_full = fat + [FATSECT] * n_fat + [DIFSECT] * n_difat
    fat_full += [FREESECT] * (n_fat * FAT_PER_SECTOR - len(fat_full))
    fat_raw = b"".join(struct.pack("<I", v) for v in fat_full)
    fat_sector_ids = list(range(n_data, n_data + n_fat))
    difat_sector_ids = list(range(n_data + n_fat, n_data + n_fat + n_difat))
    for i in range(n_fat):
        sectors.append(fat_raw[i * SECTOR:(i + 1) * SECTOR])
    for i in range(n_difat):
        chunk = fat_sector_ids[109 + i * DIFAT_PER_SECTOR: 109 + (i + 1) * DIFAT_PER_SECTOR]
        chunk += [FREESECT] * (DIFAT_PER_SECTOR - len(chunk))
        nxt = difat_sector_ids[i + 1] if i + 1 < n_difat else ENDOFCHAIN
        sectors.append(b"".join(struct.pack("<I", v) for v in chunk) + struct.pack("<I", nxt))

    # ---- header ------------------------------------------------------------
    hdr = bytearray(SECTOR)
    struct.pack_into("<8s16sHHHHH6sIIIIIIIII", hdr, 0,
                     b"\xd0\xcf\x11\xe0\xa1\xb1\x1a\xe1", b"\0" * 16,
                     0x003E, 0x0003, 0xFFFE, 9, 6, b"\0" * 6,
                     0,               # number of directory sectors (v3: 0)
                     n_fat,
                     dir_start,
                     0,               # transaction signature
                     MINI_CUTOFF,
                     minifat_start, minifat_count,
                     difat_sector_ids[0] if n_difat else ENDOFCHAIN, n_difat)
    difat = fat_sector_ids[:109] + [FREESECT] * max(0, 109 - len(fat_sector_ids))
    struct.pack_into("<109I", hdr, 76, *difat)

    return bytes(hdr) + b"".join(sectors)


def _dir_entry(e: Optional[_Entry]) -> bytes:
    if e is None:
        return struct.pack("<64sHBBIII16sIQQIQ", b"\0" * 64, 0, 0, 0,
                           NOSTREAM, NOSTREAM, NOSTREAM, b"\0" * 16, 0, 0, 0, 0, 0)
    name = e.name.encode("utf-16-le") + b"\0\0"
    if len(name) > 64:
        raise ValueError(f"name too long: {e.name!r}")
    return struct.pack("<64sHBBIII16sIQQIQ",
                       name.ljust(64, b"\0"), len(name), e.etype, 1,  # colour: black
                       e.left, e.right, e.child, e.clsid, 0, 0, 0,
                       e.start if e.start != ENDOFCHAIN else (0 if e.etype != TYPE_STREAM else ENDOFCHAIN),
                       e.size)


def load_cfb(path: str) -> Storage:
    """Read an existing compound file into a Storage tree (via olefile)."""
    import olefile
    ole = olefile.OleFileIO(path)
    root = Storage()
    for parts in ole.listdir(streams=True, storages=True):
        p = "/".join(parts)
        if ole.get_type(p) == olefile.STGTY_STORAGE:
            root.storage_path(p)
        else:
            root.set_path(p, ole.openstream(p).read())
    return root
